tests/zquantum/core/trackers_test.py

```diff
@@ -0,0 +1,146 @@
+import json
+from functools import partial
+from os import remove
+
+import networkx as nx
+import numpy as np
+import pytest
+from qeqiskit.simulator import QiskitSimulator
+from zquantum.core.cost_function import AnsatzBasedCostFunction
+from zquantum.core.estimation import (
+    allocate_shots_uniformly,
+    estimate_expectation_values_by_averaging,
+)
+from zquantum.core.trackers import MeasurementTrackingBackend
+from zquantum.optimizers.scipy_optimizer import ScipyOptimizer
+from zquantum.qaoa.ansatzes.farhi_ansatz import QAOAFarhiAnsatz
+from zquantum.qaoa.problems.maxcut import MaxCut
+
+
+@pytest.fixture
+def H():
+    G = nx.Graph()
+    G.add_nodes_from([0, 1, 2, 3])
+    G.add_edge(0, 1, weight=10)
+    G.add_edge(0, 3, weight=10)
+    G.add_edge(1, 2, weight=1)
+    G.add_edge(2, 3, weight=1)
+    problem = MaxCut()
+    return problem.get_hamiltonian(G)
+
+
+class TestMeasurementTrackingBackend:
+    def test_save_measurement_data_from_maxcut_qaoa(self, H):
+        # Given
+        ansatz = QAOAFarhiAnsatz(1, cost_hamiltonian=H)
+        backend = MeasurementTrackingBackend(
+            inner_backend=QiskitSimulator(device_name="aer_simulator")
+        )
+
+        estimation_method = estimate_expectation_values_by_averaging
+        shot_allocation = partial(allocate_shots_uniformly, number_of_shots=100)
+        estimation_preprocessors = [shot_allocation]
+        optimizer = ScipyOptimizer(method="L-BFGS-B", options={"maxiter": 10})
+        cost_function = AnsatzBasedCostFunction(
+            H, ansatz, backend, estimation_method, estimation_preprocessors
+        )
+        initial_params = np.array([0, 0])
+        # When
+        opt_results = optimizer.minimize(cost_function, initial_params)
+        circuit = ansatz.get_executable_circuit(opt_results.opt_params)
+        backend.n_samples = 10000
+        backend.run_circuit_and_measure(circuit, n_samples=1000)
+        f = open(backend.file_name)
```

Comment:
1. If you don't wrap `open(...)` in a context manager you're responsible for closing the file pointer on your own. I can't see the `f.close()` in this test, so I'd expect FS errors due to removal of an dangling/opened file at the test cleanup. Is there a reason not to use `with open(...) as f`?
2. If this test fails for some reason we end up with temporary files cluttering the current PWD. WDYT about using a [tempfile]()https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) or a [tempdir](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory)? You can see how we use that in [SDK's tests](https://github.com/zapatacomputing/orquestra-sdk/blob/a9b2f1d94d0f7671c7bbb357acfde1ffb8dbb9ba/tests/v2/test_dispatch.py#L10-L46)